python/lsst/sims/maf/runComparison/runComparison.py

```diff
@@ -542,3 +627,176 @@ def plotMetricData(self, bundleDict, plotFunc, runlist=None, userPlotDict=None,
         if self.verbose:
             print(plotDicts)
         ph.plot(plotFunc, plotDicts=plotDicts)
+
+    def generateDiffHtml(self, normalized = False, html_out = None, show_page = False,
+                         combined = False, fullStats = False):
+        """
+        Use `bokeh` to convert a summaryStats dataframe to interactive html
+        table.
+
+        Parameters
+        ----------
+        normalized : bool, opt
+            If True generate html table with normalizedStats
+        html_out : str, opt
+            Name of the html that will be output and saved. If no string
+            is provided then the html table will not be saved.
+        show_page : bool, opt
+            If True the html page generate by this function will automatically open
+            in your browser
+        combined : bool, opt
+            If True the html produce will have columns for the original
+            summaryStats values, as well as their normalized values. The baselineRun
+            used to calculate the normalized values will be dropped from the table.
+        fullStats : bool, opt
+            If False the final html table will not include summaryStats that
+            contain '3Sigma','Rms','Min','Max','RobustRms', or '%ile' in their
+            names.
+        """
+
+        try:
+            import bokeh
+            BOKEH_HERE = True
+        except ImportError:
+            BOKEH_HERE = False
+            raise ImportError('This method requires bokeh to be installed.'+ '\n'
+                              'Run: pip install bokeh'+'\n' +
+                              'Then restart your jupyter notebook kernel.')
+
+        if BOKEH_HERE is True:
```

Comment:
I agree that moving the `bokeh` import to the beginning of the script would be better style wise. Although for now we wanted to isolate the things using `bokeh` to the `generareDiffHtml` method so that not having `bokeh` installed would not affect the rest of the `runComparrison` class. 

To skip the if statement at line `666` I will move the rest of the `bokeh` imports under the `try` at line `657`. Is that what you had in mind?

Something like:
~~~
try:
    from bokeh.models import CustomJS, ColumnDataSource
    from bokeh.io import output_file, output_notebook
    from bokeh.layouts import widgetbox, layout, row, column
    from bokeh.models.widgets import DataTable, DateFormatter, TableColumn, 
    NumberFormatter, Select
    from bokeh.plotting import Figure, output_file, show
    output_notebook()
except ImportError:
    raise ImportError('This method requires bokeh to be installed.'+ '\n'
                                 'Run: pip install bokeh'+'\n' +
                                 'Then restart your jupyter notebook kernel.')
~~~

Reply:
That is more in line with the style adopted there yes. But note that you could keep the same functionality by moving the try/expect to the head of the script and using `BOKEH_HERE` as a global variable. Something like:

```
_BOKEH_HERE = True
try:
    from bokeh.models import CustomJS, ColumnDataSource
    from bokeh.io import output_file, output_notebook
    from bokeh.layouts import widgetbox, layout, row, column
    from bokeh.models.widgets import DataTable, DateFormatter, TableColumn, 
    NumberFormatter, Select
    from bokeh.plotting import Figure, output_file, show
    output_notebook()
except ImportError:
    _BOKEH_HERE = False
    warnings.warn('This method requires bokeh to be installed.'+ '\n'
                                 'Run: pip install bokeh'+'\n' +
                                 'Then restart your jupyter notebook kernel.')
```

Then, on the function you could use

```
if not _BOKEH_HERE:
    raise ImportError('This method requires bokeh to be installed.'+ '\n'
                                 'Run: pip install bokeh'+'\n' +
                                 'Then restart your jupyter notebook kernel.')

```

But honestly, no strong opinion agains one or the other. I am not a big fan of having imports inside functions but unless there's a guideline that specifies you shouldn't do it then, why not?

Reply:
Ah, I see what you mean now. I just tried this out and it is a much cleaner way of doing it. I will push a new commit. Thanks.

Reply:
Yeap, looks much better... 😃  👍